src/init.cpp

```diff
@@ -793,6 +793,32 @@ void InitParameterInteraction(ArgsManager& args)
             LogInfo("parameter interaction: -onlynet excludes IPv4 and IPv6 -> setting -dnsseed=0\n");
         }
     }
+
+    // If settings.json contains a "upnp" option, migrate it to use "natpmp" instead
+    bool settings_changed{false}; // Whether settings.json file needs to be rewritten
+    args.LockSettings([&](common::Settings& settings) {
+        if (auto* upnp{common::FindKey(settings.rw_settings, "upnp")}) {
+            if (common::FindKey(settings.rw_settings, "natpmp") == nullptr) {
+                LogWarning(R"(Adding "natpmp": %s to settings.json to replace obsolete "upnp" setting)", upnp->write());
+                settings.rw_settings["natpmp"] = *upnp;
+            }
+            LogWarning(R"(Removing obsolete "upnp" setting from settings.json)");
+            settings.rw_settings.erase("upnp");
+            settings_changed = true;
+        }
+    });
+    if (settings_changed) args.WriteSettingsFile();
```

Comment:
Note: `WriteSettingsFile()` throws if the `-nosettings` argument is set, but in that case we would never find a value for `upnp` above, and `settings_changed` won't be true, so that edge case is handled well.

Reply:
Yes, i used to have an explicit check around this logic for dynamic settings to be enabled (so that it won't throw), but removed it after a comment by @ryanofsky. It indeed doesn't seem to be necessary because the dynamic settings won't be populated in the first place.

Reply:
Thanks, I missed that discussion. (https://github.com/bitcoin/bitcoin/pull/31916#discussion_r1965808053)